# Using the SEED Kmer identification tools

In SEED we have two primary mechanisms for propagating function from the signature
kmer databases to input sequences.

## Kmer calling web service

For production use for processing protein sequences (and for making PATtyfam assignments) we
have a web service available inside on the SEED systems within the ANL firewall. The endpoint for 
the kmer function calling is `http://pear:6100/query`. The protein sequence is passed as the input 
to the HTTP POST request. For example, if the file query.fa holds one or more protein sequences in FASTA format
the curl command may be used to process the data:

```
$ curl --data-binary @query.fa http://pear:6100/query
PROTEIN-ID	     id-83333.1.peg.1	21
CALL		     0			19	13	30895	Thr operon leader peptide	22.729
OTU-COUNTS	     id-83333.1.peg.1[21]	1-91	1-200	1-251	   1-460  1-475
PROTEIN-ID	     id-83333.1.peg.2		820
CALL		     0				818	791	3449	Aspartokinase (EC 2.7.2.4) / Homoserine dehydrogenase (EC 1.1.1.3)	2432.37
OTU-COUNTS	     id-83333.1.peg.2[820]	8-310	7-398	6-163	6-244	      5-11
PROTEIN-ID	     id-83333.1.peg.3		310
CALL		     0				307	293	15456	Homoserine kinase (EC 2.7.1.39)	804.771
OTU-COUNTS	     id-83333.1.peg.3[310]	4-907	4-208	3-189	3-695	   3-378
PROTEIN-ID	     id-83333.1.peg.4		428
CALL		     0				426	412	30907	Threonine synthase (EC 4.2.3.1)	1232.38
OTU-COUNTS	     id-83333.1.peg.4[428]	8-310	5-438	5-623	4-3	  4-12
```

The fields in the output are as follows.

 * PROTEIN-ID: identifier (from fasta file), sequence length

 * CALL: start of called region (in the protein sequence), end of called region, count of kmer hits

 * OTU-COUNTS: sequence id, sequence length (in brackets), and a list of otu signatures of the format Count-OTU

If you wish to see a compact result which is generated by processing the possible multiple hits into a single call
deemed best, use the endpoint `http://pear:6100/query?find_best_call=1`:

```
$ curl --data-binary @query.fa "http://pear:6100/query?find_best_call=1"
id-83333.1.peg.1	Thr operon leader peptide	13	22.729
id-83333.1.peg.2	Aspartokinase (EC 2.7.2.4) / Homoserine dehydrogenase (EC 1.1.1.3)	791	2432.37
id-83333.1.peg.3	Homoserine kinase (EC 2.7.1.39)		293	      804.771
id-83333.1.peg.4	Threonine synthase (EC 4.2.3.1)		412	      1232.38
```

Here, the fields are sequence id, best call, score, weighted score.

## `kmer_search` command line tool

In order to access arbitrary kmer builds, or to process DNA input, the command line tool `kmer_search` found 
in the SEED codebase can be used. This tool requires a kmer data directory to be specified. The production version used 
in RAST and BV-BRC (and in the web service described above) is available in `/vol/core-seed/kmers/core.2018-0531/Data.2`. A newer
version built this month (April, 2022) is available in `/vol/core-seed/kmers/core.2022-0406/Data.2`.

